### PR TITLE
Mount concurrent read

### PR DESCRIPTION
**Pull request by zemul**
# What problem are we solving?
#4350

# How are we solving the problem?
use RWMutex.

# How is the PR tested?

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

**Comment by zemul** [comment]
before
```
Jobs: 16 (f=16): [r(16)][100.0%][r=8176KiB/s,w=0KiB/s][r=2044,w=0 IOPS][eta 00m:00s]
rand-read: (groupid=0, jobs=16): err= 0: pid=17728: Wed Mar 29 14:41:20 2023
   read: IOPS=2432, BW=9729KiB/s (9963kB/s)(570MiB/60008msec)
    slat (usec): min=620, max=26847, avg=6568.13, stdev=782.88
    clat (nsec): min=952, max=45393, avg=3519.05, stdev=3391.38
     lat (usec): min=623, max=26849, avg=6572.80, stdev=782.99
    clat percentiles (nsec):
     |  1.00th=[ 1160],  5.00th=[ 1400], 10.00th=[ 1656], 20.00th=[ 1976],
     | 30.00th=[ 2256], 40.00th=[ 2384], 50.00th=[ 2416], 60.00th=[ 2736],
     | 70.00th=[ 3440], 80.00th=[ 4128], 90.00th=[ 4960], 95.00th=[ 5920],
     | 99.00th=[18816], 99.50th=[19328], 99.90th=[25984], 99.95th=[30080],
     | 99.99th=[34048]
   bw (  KiB/s): min=  504, max=  648, per=6.25%, avg=608.00, stdev=25.49, samples=1920
   iops        : min=  126, max=  162, avg=151.98, stdev= 6.38, samples=1920
  lat (nsec)   : 1000=0.01%
  lat (usec)   : 2=20.87%, 4=56.67%, 10=17.95%, 20=4.31%, 50=0.20%
  cpu          : usr=0.16%, sys=0.32%, ctx=146098, majf=0, minf=553
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=145955,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: bw=9729KiB/s (9963kB/s), 9729KiB/s-9729KiB/s (9963kB/s-9963kB/s), io=570MiB (598MB), run=60008-60008msec
```
afer
```
rand-read: (groupid=0, jobs=16): err= 0: pid=21695: Thu Apr 13 16:18:50 2023
   read: IOPS=25.3k, BW=98.7MiB/s (103MB/s)(5921MiB/60001msec)
    clat (usec): min=259, max=17199, avg=631.54, stdev=175.14
     lat (usec): min=259, max=17199, avg=631.70, stdev=175.14
    clat percentiles (usec):
     |  1.00th=[  388],  5.00th=[  449], 10.00th=[  482], 20.00th=[  523],
     | 30.00th=[  562], 40.00th=[  586], 50.00th=[  611], 60.00th=[  635],
     | 70.00th=[  668], 80.00th=[  709], 90.00th=[  775], 95.00th=[  857],
     | 99.00th=[ 1254], 99.50th=[ 1500], 99.90th=[ 2180], 99.95th=[ 2474],
     | 99.99th=[ 3556]
   bw (  KiB/s): min= 5688, max= 6528, per=6.25%, avg=6314.74, stdev=69.02, samples=1910
   iops        : min= 1422, max= 1632, avg=1578.66, stdev=17.26, samples=1910
  lat (usec)   : 500=14.04%, 750=72.99%, 1000=10.62%
  lat (msec)   : 2=2.21%, 4=0.14%, 10=0.01%, 20=0.01%
  cpu          : usr=0.50%, sys=1.20%, ctx=1516583, majf=0, minf=1163
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=1515669,0,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: bw=98.7MiB/s (103MB/s), 98.7MiB/s-98.7MiB/s (103MB/s-103MB/s), io=5921MiB (6208MB), run=60001-60001msec
```

File: weed/filer/filechunk_section.go [hunk 1]
```diff
@@ -13,7 +13,8 @@ type FileChunkSection struct {
 	visibleIntervals *IntervalList[*VisibleInterval]
 	chunkViews       *IntervalList[*ChunkView]
 	reader           *ChunkReadAt
-	lock             sync.Mutex
+	lock             sync.RWMutex
+	prepare          bool
```
Comment (chrislusf) [inline weed/filer/filechunk_section.go]: change to "isPrepared"?

**Comment by chrislusf** [comment]
The speed improvement is too good to believe. :)

File: weed/filer/filechunk_section.go (hunk 1 above)
Comment (zemul) [reply weed/filer/filechunk_section.go]: > change to "isPrepared"?

done.

Comment (lizhengui007) [reply weed/filer/filechunk_section.go]: what is the "Prepared" used for ?

Comment (zemul) [reply weed/filer/filechunk_section.go]: > what is the "Prepared" used for ?

Avoid adding write lock to section during concurrent read.

Comment (lizhengui007) [reply weed/filer/filechunk_section.go]: Currently, file openings share the same file handle structure. Perhaps it is a better design to get an independent file handle structure for opening each time.

**Review by chrislusf — COMMENTED** [review]
this double checking is necessary since the locked section will change the isPrepared value?

**Review by chrislusf — CHANGES_REQUESTED** [review]
revert this?

File: weed/filer/filechunk_section.go (hunk 1 above)
Comment (zemul) [reply weed/filer/filechunk_section.go]: > Currently, file openings share the same file handle structure. Perhaps it is a better design to get an independent file handle structure for opening each time.

This is better,  read and write consistency needs to be considered.

**Comment by KyleSanderson** [comment]
> Thanks a lot! The speed improvement is too good to believe. :) I need to test it myself also.

how'd it go?

File: weed/filer/filechunk_section.go [hunk 2]
```diff
@@ -61,6 +62,19 @@ func removeGarbageChunks(section *FileChunkSection, garbageFileIds map[string]st
 }
 
 func (section *FileChunkSection) setupForRead(group *ChunkGroup, fileSize int64) {
+	if section.isPrepared {
+		section.reader.fileSize = fileSize
+		return
+	}
```
Comment (half-life666) [inline weed/filer/filechunk_section.go:68]: Why 65-68 is same with 73-76? Is it intentional or mistake?

Because 65-68 is not thread safe.

**Pull request merged**